### ISSUE TITLE
Better handling for outband power operations on VR

### ISSUE THREAD

**Issue by rajujith** [apache/cloudstack#8284]
##### ISSUE TYPE

 * Bug Report
 * Improvement Request
 

##### COMPONENT NAME

~~~
VR,VMware
~~~

##### CLOUDSTACK VERSION

~~~
4.17.2
~~~

##### SUMMARY

When a VR is stopped on cloudstack then management/control IP is released. 

If the VR is powered on later on the hypervisor the VR will function normally with its previous management IP but CloudStack won’t find this IP information although the power state is synced. 

This situation could result in operations issues on the guest works since either : 

    There is a management IP on the VR but cloudstack doesn’t know about it so most of the operations can fail. 

    The previously released IP could be allocated to a new VR resulting in IP conflict and potentially configuring the wrong VR with network configurations as well. 

##### STEPS TO REPRODUCE

~~~
    1.Deploy a VM on a new isolated guest network with a default offering.

    2.Stop the VR in cloudStack

    3. Start the VR on the hypervisor. 
~~~

A possible solution is to restart the VR once it detects an out-of-band power on.

**Comment by rajujith** [comment]
@weizhouapache  there could be use cases where the VR would be started outside the proposed reservation time, in such cases it is still good to restart the VR to avoid those corner cases too.

**Comment by DaanHoogland** [comment]
Good find @weizhouapache however, I think as the router is started out of bounds, and hance was down, we shouldn't worry to much about downtime. The solution looks fine but over complicated to me. I think we can in genera just reboot (if that adequately takes care of all possible IP conflicts.

**Comment by andrijapanicsb** [comment]
Guys, we here have a DESIGN issue - and (IMO) nothing else. Due to the design issue - we are seeing a bug in certain situations.

CloudStack should NOT be marking the management IP of the VR as free in the DB when a VR is stopped (which is what happens when you get PowerOff report from VMware during out-of-band VR operations - stopping a VR in vSphere, DRS moving VR, or even manual live migrations in vSphere).

If we don't release the IP in the DB (nor i.e. remove nics from the nics table) - there will be no bug/issue present..

It makes zero sense to remove IPs in DB when you stop the VM - the VM is still there, why remove it's IP/nic in DB?
Unless there is a very specific reason that we are unaware of.

**Comment by alexandremattioli** [comment]
Suggestion.  
New global setting "release management IP"  true/false.

If true then behaviour is as is.
If false then ACS doesn't release the IP.

@DaanHoogland @weizhouapache @NuxRo @rajujith @andrijapanicsb

**Comment by DaanHoogland** [comment]
> Suggestion. New global setting "release management IP" true/false.
> 
> If true then behaviour is as is. If false then ACS doesn't release the IP.

I would add that the new behaviour will be default behaviour for new installations and the old behaviour will be the behavour on upgrades (if feasible)

> 
> @DaanHoogland @weizhouapache @NuxRo @rajujith @andrijapanicsb

**Comment by andrijapanicsb** [comment]
I'm against this **"the old behaviour will be the behaviour on upgrades (if feasible)"** - because then our fix - fixes nothing for existing customers.

I'm of the opinion that we change new installations/upgrades to the new behaviour, but consider putting an explanation in the documentation PR (not sure if anyone will care, but for completeness).

**Comment by weizhouapache** [comment]
let's use the same behaviour for same cloudstack versions, no matter it is fresh installation or upgraded from older versions.
the behaviour is determined by the new global setting.

**Comment by DaanHoogland** [comment]
> I'm against this **"the old behaviour will be the behaviour on upgrades (if feasible)"** - because then our fix - fixes nothing for existing customers.

That is not true. They have to take action setting the setting, but we won't be sabotaging working environments that would not want this.

**Comment by weizhouapache** [comment]
> > I'm against this **"the old behaviour will be the behaviour on upgrades (if feasible)"** - because then our fix - fixes nothing for existing customers.
> 
> That is not true. They have to take action setting the setting, but we won't be sabotaging working environments that would not want this.

@DaanHoogland 
I am not sure how large the impact is. 

Some users might encounter the issue that the pod management IPs get consumed more than they expected. If all pod management IPs are used, they need to update or add the management IPs by API
https://cloudstack.apache.org/api/apidocs-4.18/apis/updatePodManagementNetworkIpRange.html
https://cloudstack.apache.org/api/apidocs-4.18/apis/createManagementNetworkIpRange.html
https://cloudstack.apache.org/api/apidocs-4.18/apis/deleteManagementNetworkIpRange.html

All the management IPs must be in same subnet (same gateway/netmask). It might be a major issue.
```
        // Because each pod has only one Gateway and Netmask.
        if (!gateway.equals(pod.getGateway())) {
            throw new InvalidParameterValueException("Multiple gateways for the POD: " + pod.getId() + " are not allowed. The Gateway should be same as the existing Gateway " + pod.getGateway());
        }

        if (!netmask.equals(NetUtils.getCidrNetmask(cidrSize))) {
            throw new InvalidParameterValueException("Multiple subnets for the POD: " + pod.getId() + " are not allowed. The Netmask should be same as the existing Netmask " + NetUtils.getCidrNetmask(cidrSize));
        }
```

**Comment by andrijapanicsb** [comment]
By removing NIC IPs of the VR, you give the false expectation that they have enough IPs - if they need to start VRs that are stopped (but IPs are now being used for other VRs) - i.e. if they have more VRs deployed then the number of MGMT IPs for the VRs - they will hit a hard wall with customers not being able to power on their VMs (networks, due to VR can't start).

That is much worse than "exhausting" all mgmt IPs assigned and only failing to deploy NEW networks (vs. not being able to "start" old networks/VMs).

**Comment by andrijapanicsb** [comment]
(in other words you are speculating that some people might use it, while in reality we had known users hit the wall with the bug)
(i.e. let's not speculate?)

**Comment by DaanHoogland** [comment]
The idea is that you don't run all at once and saying people don´t use it like it was designed is as much speculating as saying that they do.

**Comment by andrijapanicsb** [comment]
You want to allow normal users to hit issues, while those "abusing" and misconfiguring their mgmt IPs should still continue to do so and eventually (also) shoot themself in the foot? 

OK, we have made MUCH worse decisions in the past - be my guest, do it as you guys prefer. Cheers

**Comment by weizhouapache** [comment]
> You want to allow normal users to hit issues, while those "abusing" and misconfiguring their mgmt IPs should still continue to do so and eventually (also) shoot themself in the foot?
> 
> OK, we have made MUCH worse decisions in the past - be my guest, do it as you guys prefer. Cheers

I am OK if the private IPs are not released by default.
If the users do not have free private IPs, they can remove Stopped VRs which will release the private IPs.

**Comment by DaanHoogland** [comment]
as remarked in https://github.com/apache/cloudstack/pull/8371#issuecomment-1865977414 the oob stopped VR also causes the guest vlan to be removed from the VR guest nic

**Comment by weizhouapache** [comment]
fixed by #8284
